leap_c/examples/hvac/controller.py

```diff
@@ -0,0 +1,542 @@
+    solar_forecast = solar_forecast.reshape(-1)
+    price_forecast = price_forecast.reshape(-1)
+    time = time.reshape(-1)
+    
+    quarter_hours = np.arange(obs[0], obs[0] + len(time)) % len(time)
+    T_lower, T_upper = set_temperature_limits(quarter_hours=quarter_hours)
+    
+    T_lower_celsius = convert_temperature(T_lower.reshape(-1), "kelvin", "celsius")
+    T_upper_celsius = convert_temperature(T_upper.reshape(-1), "kelvin", "celsius")
+    Ta_celsius = convert_temperature(Ta_forecast.reshape(-1), "kelvin", "celsius")
+    
+    # Create base plot
+    fig, axes = _create_base_plot(figsize)
+    
+    # Plot each subplot using helper functions
+    _plot_temperature_subplot(axes[0], time, Ti_celsius, Te_celsius, T_lower_celsius, T_upper_celsius)
+    _plot_heater_subplot(axes[1], time, Th_celsius)
+    _plot_disturbance_subplot(axes[2], time, Ta_celsius, solar_forecast)
+    _plot_control_subplot(axes[3], time, u, price_forecast)
+    
+    # Adjust layout and add summary stats
+    plt.tight_layout()
+    _add_summary_stats(fig, u, ctx)
+    
+    # Save figure if path provided
+    if save_path:
+        plt.savefig(save_path, dpi=300, bbox_inches="tight")
+        print(f"Figure saved to: {save_path}")
+    
+    return fig
+
+
+def plot_simulation(time: np.ndarray, obs: np.ndarray, action: np.ndarray) -> plt.Figure:
+    """
+    Plot simulation results in a figure with four vertically stacked subplots.
+
+    Args:
+        time: Time array
+        obs: Observation data
+        action: Control action array
+
+    Returns:
+        matplotlib Figure object
+    """
+    quarter_hours, day, Ti, Th, Te, Ta_forecast, solar_forecast, price_forecast = decompose_observation(obs=obs)
+    
+    # Convert temperatures to Celsius for plotting
+    Ti_celsius = convert_temperature(Ti, "kelvin", "celsius")
+    Th_celsius = convert_temperature(Th, "kelvin", "celsius")
+    Te_celsius = convert_temperature(Te, "kelvin", "celsius")
+    Ta_celsius = convert_temperature(Ta_forecast[:, 0], "kelvin", "celsius")
+    
+    Ti_lower, Ti_upper = set_temperature_limits(quarter_hours)
+    Ti_lower_celsius = convert_temperature(Ti_lower, "kelvin", "celsius")
+    Ti_upper_celsius = convert_temperature(Ti_upper, "kelvin", "celsius")
+    
+    qh = action
+    solar = solar_forecast[:, 0]
+    price = price_forecast[:, 0]
+    
+    # Create base plot
+    fig, axes = _create_base_plot()
+    
+    # Plot each subplot using helper functions
+    _plot_temperature_subplot(axes[0], time, Ti_celsius, Te_celsius, Ti_lower_celsius, Ti_upper_celsius)
+    _plot_heater_subplot(axes[1], time, Th_celsius)
+    _plot_disturbance_subplot(axes[2], time, Ta_celsius, solar)
+    _plot_control_subplot(axes[3], time, qh, price)
+    
+    # Adjust layout and add summary stats
+    plt.tight_layout()
+    _add_summary_stats(fig, qh)
+    
+    return fig
+
+
+if __name__ == "__main__":
+    horizon_hours = 24
```

Comment:
maybe migrate this to some `main.py` file or a script?
Ideally also test it on actions.